### ISSUE TITLE
Running planemo workflow_lint on changed workflows

### ISSUE THREAD

**Issue by hexylena** [galaxyproject/training-material#3970]
> Please make sure that "workflow_outputs" for every step have labels.
> The "best practice" wizard in the WF editor should also warn about this.

_Originally posted by @wm75 in https://github.com/galaxyproject/training-material/pull/3958#discussion_r1147394575_

**Comment by hexylena** [comment]
We discussed out of band, how to achieve this. Planemo itself is far, far too slow to be integrated into our review process:

- Every workflow takes ~10 seconds to check with the toolshed (which we can eliminate with `--skip tool_ids`)
- Another 1 second just to load modules from disk (somewhat amortised with linting multiple workflows in one go)
- but still like 200-300ms per WF which is very high for us

We can try and work around this by only linting changed workflows but things just shouldn't be that slow. 

So we can _either_ re-implement the checks in a more lightweight manner (ugh, but it's just a json file, we could achieve speed for the 95% use case and lint on the major issues.) _or_ we can filter for changed files only (so the linting only works in github actions, sad.)

**Comment by wm75** [comment]
Re-implementing checks is not unheard of: @davelopez has done sth like this in typescript before for his https://github.com/davelopez/galaxy-workflows-vscode.

Just like there, you would have to repeat the json exercise on workflow format 2 though at some point.

**Comment by davelopez** [comment]
At some point, I want to extract both language servers `.ga (JSON)` and `.yml` on their own npm packages, and maybe it makes sense to extract basic lightweight Typescript libraries from them to reuse in other frontend projects if that is useful.

**Comment by hexylena** [comment]
yeah, I think it might be an acceptable tradeoff. The current tests that we get warned on look like:

Error Message | Can we Check It
--- | ---
.. CHECK: Tests appear structurally correct for topics/metagenomics/tutorial | No, we'd lose this
.. WARNING: Input X of workflow step 0 is disconnected. | I'd rather not implement this
.. WARNING: Workflow contained no outputs | easy
.. WARNING: Workflow contained output without a label | easy
.. WARNING: Workflow does not specify a creator. | we already test for this
.. WARNING: Workflow missing test cases. | we already test for this
.. WARNING: Workflow step with ID X has no annotation. | easy
.. WARNING: Workflow step with ID X has no label. | easy

Kinda funny seeing it described as "legacy" and "internal only" in your vscode extension repo, as we're really only concerned with `.ga` since that is still the default download option (is it even possible to get format2 from the UI?). But yes, if that was extracted, that'd be interesting. I'd love to be able to run your simplified diff on workflow changes, that'd save a lot of time reviewing.

**Comment by hexylena** [comment]
But this is kinda a "whenever it happens it happens" project, it's not needed on a short timeline

**Comment by davelopez** [comment]
> Kinda funny seeing it described as "legacy" and "internal only" in your vscode extension repo

I thought that too, but what I learned is the JSON format was never meant (or designed) to be "editable by hand", the extension supporting it was just a way of easing the pain points in the transition period to the .yml format which I guess will be fully enabled soon (hopefully, I don't really know or remember what is still missing). Another concern was having 2 schemas, one *unofficial* for the JSON and the official `gxformat2` so people might be confused and think that the JSON schema was the way to go...

**Comment by hexylena** [comment]
the GTN will probably continue using json as the primary format until it becomes the default way to download, and even then it'll take a release or two before we know all of our contributors are updated and generating yaml by default, and probably at that time we'll batch convert
